image-loader/app/model/ImageUpload.scala

```diff
@@ -28,57 +29,92 @@ case object ImageUpload {
       case Some("image/png") => FileMetadataReader.fromICPTCHeadersWithColorInfo(uploadedFile)
-      case _                 => FileMetadataReader.fromIPTCHeaders(uploadedFile)
+      case _ => FileMetadataReader.fromIPTCHeaders(uploadedFile)
     }
 
-    val thumbFuture            = for {
-      fileMetadata   <- fileMetadataFuture
-      colourModel    <- colourModelFuture
-      iccColourSpace  = FileMetadataHelper.normalisedIccColourSpace(fileMetadata)
-      thumb          <- ImageOperations.createThumbnail(uploadedFile, thumbWidth, thumbQuality, tempDir, iccColourSpace, colourModel)
-    } yield thumb
-
-    bracket(thumbFuture)(_.delete) { thumb =>
-      // Run the operations in parallel
-      val thumbStoreFuture      = storeThumbnail(uploadRequest, thumb)
-      val thumbDimensionsFuture = FileMetadataReader.dimensions(thumb, Some("image/jpeg"))
-
-      for {
-        s3Source         <- sourceStoreFuture
-        s3Thumb          <- thumbStoreFuture
-        sourceDimensions <- sourceDimensionsFuture
-        thumbDimensions  <- thumbDimensionsFuture
-        fileMetadata     <- fileMetadataFuture
-        colourModel      <- colourModelFuture
-        fullFileMetadata  = fileMetadata.copy(colourModel = colourModel)
-
-        metadata      = ImageMetadataConverter.fromFileMetadata(fullFileMetadata)
-        cleanMetadata = metadataCleaners.clean(metadata)
-
-        sourceAsset = Asset.fromS3Object(s3Source, sourceDimensions)
-        thumbAsset  = Asset.fromS3Object(s3Thumb,  thumbDimensions)
-
-        baseImage      = createImage(uploadRequest, sourceAsset, thumbAsset, fullFileMetadata, cleanMetadata)
-        processedImage = SupplierProcessors.process(baseImage)
-
-        // FIXME: dirty hack to sync the originalUsageRights and originalMetadata as well
-        finalImage     = processedImage.copy(
-          originalMetadata    = processedImage.metadata,
-          originalUsageRights = processedImage.usageRights
-        )
-      }
-      yield {
-        ImageUpload(uploadRequest, finalImage)
+    for {
+      fileMetadata <- fileMetadataFuture
+    } yield {
+
+      val uploadIsPng24 = isPng24(uploadRequest.mimeType, fileMetadata)
+
+      // These futures are started outside the for-comprehension, otherwise they will not run in parallel
+      val sourceStoreFuture = storeSource(uploadRequest)
+      // FIXME: pass mimeType
+      val colourModelFuture = ImageOperations.identifyColourModel(uploadedFile, "image/jpeg")
+      val sourceDimensionsFuture = FileMetadataReader.dimensions(uploadedFile, uploadRequest.mimeType)
+
+      val thumbFuture = for {
+        fileMetadata <- fileMetadataFuture
+        colourModel <- colourModelFuture
+        iccColourSpace = FileMetadataHelper.normalisedIccColourSpace(fileMetadata)
+        thumb <- ImageOperations.createThumbnail(uploadedFile, thumbWidth, thumbQuality, tempDir, iccColourSpace, colourModel)
+      } yield thumb
+
+      val pngOption = if (uploadIsPng24) {
```

Comment:
I was more thinking about encapsulating the `isPng` check within an object. Something like:

```scala
case class PngThing(optimisedFile: Option[File], dimensions: Future[Option[Dimensions]])

case object PngThing {
  private def isPng(file: File) = true
  
  def build (file: File) = {
    if (isPng(file)) {
      val optimisedFile = {
        val optimsedFile = tempDir.getAbsolutePath() + "/optimisedpng" + ".png"
        Seq("pngquant", "--quality", "1-85", file.getAbsolutePath(), "--output", optimsedFile)
        new File(optimsedFile)
      }

      val dimensions: Future[Option[Dimensions]] = FileMetadataReader.dimensions(optimisedFile, Some("image/png"))
      
      PngThing(Some(optimisedFile), dimensions)
    }
    else {
      PngThing(None, Future(None))
    }
  }
}

```

then usage becomes

```scala
val pngThing = PngThing.build(file)
pngThing.optimised
```

not a blocker, just a thought 😄